### PR TITLE
[CELEBORN-1579] Fix the memory leak of result partition

### PR DESCRIPTION
**Pull request by reswqa**
### What changes were proposed in this pull request?
Don't register `RemoteShuffleResultPartition` to partition manager because we don't store resource in TM.

In theory, we could also fix this on the Flink side by no longer registering partitions to the partition manager for remote/cluster partitions, but that will only be released after Flink 2.0 at the soon.

### Why are the changes needed?
`RemoteShuffleResultPartition` will be registered to the partition manager(at setup phase). Since it's a cluster partition(resources are not stored on the Flink TM), Flink does not trigger the resource releasing over TM.

In a session cluster, the partition object is leaked. As a more serious consequence, the failure of the partition to release will result in the idle TM not being reclaimed by the Flink resource manager.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Self tested in a session cluster.

### CODE REVIEW

**Comment by mridulm** [comment]
+CC @venkata91

File: client-flink/flink-1.14/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleResultPartition.java [hunk 1]
```diff
@@ -82,11 +87,17 @@ public RemoteShuffleResultPartition(
     delegation =
         new RemoteShuffleResultPartitionDelegation(
             networkBufferSize, outputGate, this::updateStatistics, numSubpartitions);
+    this.bufferPoolFactory = bufferPoolFactory;
   }
 
   @Override
   public void setup() throws IOException {
-    super.setup();
+    // We can't call the `setup` method of the base class, otherwise it will cause a partition leak.
```
Comment (SteNicholas) [inline client-flink/flink-1.14/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleResultPartition.java:95]: Could you describe why calling `setup` cause partition leak in comment?

File: client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleResultPartition.java [hunk 2]
```diff
@@ -123,7 +135,7 @@ public void abortCheckpoint(long l, CheckpointException e) {}
 
   @Override
   public void finish() throws IOException {
-    Utils.checkState(!isReleased(), "Result partition is already released.");
+    checkState(!isReleased(), "Result partition is already released.");
```
Comment (SteNicholas) [inline client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleResultPartition.java:138]: Why did you change this?

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Review by SteNicholas — APPROVED** [review]

**Review by RexXiong — APPROVED** [review]
LGTM

**Comment by SteNicholas** [comment]
Merged to main(v0.6.0) and branch-0.5(v0.5.2).
